src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -426,6 +402,115 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
+    let implicit_impls = crate::clean::get_auto_trait_and_blanket_impls(cx, ty, did);
+    let mut candidates: Vec<_> = implicit_impls
+        .flat_map(|impl_outer| {
+            match impl_outer.inner {
+                ImplItem(impl_) => {
+                    debug!("considering auto or blanket impl for trait {:?}", impl_.trait_);
+                    // Give precedence to methods that were overridden
+                    if !impl_.provided_trait_methods.contains(&*item_name.as_str()) {
+                        let mut items = impl_
+                            .items
+                            .into_iter()
+                            .filter_map(|assoc| {
+                                if assoc.name.as_deref() != Some(&*item_name.as_str()) {
+                                    return None;
+                                }
+                                let kind = assoc
+                                    .inner
+                                    .as_assoc_kind()
+                                    .expect("inner items for a trait should be associated items");
+                                if kind.namespace() != ns {
+                                    return None;
+                                }
+
+                                trace!("considering associated item {:?}", assoc.inner);
+                                // We have a slight issue: normal methods come from `clean` types,
+                                // but provided methods come directly from `tcx`.
+                                // Fortunately, we don't need the whole method, we just need to know
+                                // what kind of associated item it is.
+                                Some((assoc.def_id, kind))
+                            });
+                        let assoc = items.next();
+                        debug_assert_eq!(items.count(), 0);
+                        assoc
+                    } else {
+                        // These are provided methods or default types:
+                        // ```
+                        // trait T {
+                        //   type A = usize;
+                        //   fn has_default() -> A { 0 }
+                        // }
+                        // ```
+                        let trait_= impl_.trait_.unwrap().def_id().unwrap();
+                        cx.tcx.associated_items(trait_)
+                            .find_by_name_and_namespace(cx.tcx, Ident::with_dummy_span(item_name), ns, trait_)
+                            .map(|assoc| (assoc.def_id, assoc.kind))
+                    }
+                }
+                _ => panic!("get_impls returned something that wasn't an impl"),
+            }
+        })
+        .collect();
+    // Next consider explicit impls: `impl MyTrait for MyType`
+    // There isn't a cheap way to do this. Just look at every trait!
+    for &trait_ in cx.tcx.all_traits(LOCAL_CRATE) {
```

Comment:
Perhaps we should, at the beginning of the pass, compute this _once_ and construct a hashmap of assoc item names to trait ids.

Reply:
probably a `HashMap<str, SmallVec<[TraitId; 1]>>` or something

Reply:
In that case I'd recommend doing the specific impl lookup before blanket impls.

Reply:
At that point I'd rather make it a query so other people can reuse it. That also would get us caching for free.

Reply:
For the record - the query I want to add is `traits_implemented_by`, which avoids looking at every trait in the crate over and over again

Reply:
Added those queries. Maybe @eddyb is interested in reviewing that part of things?

Reply:
This is no longer implemented by a query because it's not stable across crates (https://github.com/rust-lang/rust/pull/74489#issuecomment-667574564). Instead rustdoc caches it for a single process execution: https://github.com/rust-lang/rust/pull/74489/commits/cbb49de8562187053573729e09d25b425e96ce37